go/vt/vttablet/onlineddl/analysis.go

```diff
@@ -0,0 +1,209 @@
+	}
+	spec := alterTable.PartitionSpec
+	if spec == nil {
+		return nil, nil
+	}
+	if spec.Action != sqlparser.DropAction {
+		return nil, nil
+	}
+	if len(spec.Names) != 1 {
+		return nil, vterrors.Errorf(vtrpcpb.Code_FAILED_PRECONDITION, "vitess only supports dropping a single partition per query: %v", sqlparser.CanonicalString(alterTable))
+	}
+	partitionName := spec.Names[0].String()
+	// OK then!
+
+	// Now, is this query dropping the first parititon in a RANGE partitioned table?
+	part := createTable.TableSpec.PartitionOption
+	if part.Type != sqlparser.RangeType {
+		return nil, nil
+	}
+	if len(part.Definitions) == 0 {
+		return nil, nil
+	}
+	var partitionDefinition *sqlparser.PartitionDefinition
+	var nextPartitionName string
+	for i, p := range part.Definitions {
+		if p.Name.String() == partitionName {
+			partitionDefinition = p
+			if i+1 < len(part.Definitions) {
+				nextPartitionName = part.Definitions[i+1].Name.String()
+			}
+			break
+		}
+	}
+	if partitionDefinition == nil {
+		// dropping a nonexistent partition. We'll let the "standard" migration execution flow deal with that.
+		return nil, nil
+	}
+	op := NewSpecialAlterOperation(dropRangePartitionSpecialOperation, alterTable, createTable)
+	op.SetDetail("partition_name", partitionName)
+	op.SetDetail("partition_definition", sqlparser.CanonicalString(partitionDefinition))
+	op.SetDetail("next_partition_name", nextPartitionName)
+	return op, nil
+}
+
+// analyzeDropFirstOrLastRangePartition sees if the online DDL drops the first or last partition
+//  in a range partitioned table
+func (e *Executor) analyzeAddRangePartition(alterTable *sqlparser.AlterTable, createTable *sqlparser.CreateTable) *SpecialAlterPlan {
+	// we are looking for a `ALTER TABLE <table> ADD PARTITION (PARTITION ...)` statement with nothing else
+	if len(alterTable.AlterOptions) > 0 {
+		return nil
+	}
+	if alterTable.PartitionOption != nil {
+		return nil
+	}
+	spec := alterTable.PartitionSpec
+	if spec == nil {
+		return nil
+	}
+	if spec.Action != sqlparser.AddAction {
+		return nil
+	}
+	if len(spec.Definitions) != 1 {
+		return nil
+	}
+	partitionDefinition := spec.Definitions[0]
+	partitionName := partitionDefinition.Name.String()
+	// OK then!
+
+	// Now, is this query adding a parititon in a RANGE partitioned table?
+	part := createTable.TableSpec.PartitionOption
+	if part.Type != sqlparser.RangeType {
+		return nil
+	}
+	if len(part.Definitions) == 0 {
+		return nil
+	}
+	op := NewSpecialAlterOperation(addRangePartitionSpecialOperation, alterTable, createTable)
+	op.SetDetail("partition_name", partitionName)
+	op.SetDetail("partition_definition", sqlparser.CanonicalString(partitionDefinition))
+	return op
+}
+
+// analyzeSpecialAlterScenarios checks if the given ALTER onlineDDL, and for the current state of affected table,
+// can be executed in a special way
+func (e *Executor) analyzeSpecialAlterPlan(ctx context.Context, onlineDDL *schema.OnlineDDL) (*SpecialAlterPlan, error) {
```

Comment:
Comment is wrong for the function name here.

Reply:
fixed